### PR TITLE
Support multiple different command load factor for the same client

### PR DESCRIPTION
**Pull request by saratry**
Introduce a RoutingSelector interface responsible to select the correct handler taken from a set of registered handlers.

The LoadFactorBasedSelector implements this interface, selecting the correct handler on load factor basis.

### CODE REVIEW

**Comment by saratry** [comment]
> The loadFactorBasedSelector has one disadvantage over the consistent hash used before. When there is a new handler for a command handler, it will impact the target for all routing keys (total changes and therefore the modulo value also changes). In the consistent hash implementation, the target only changed for the routing keys that will now go to the new handler.
> 
> The same applies when a handler is removed.

Yes, it's true. This in favour of greater lightness. Furthermore, not even in the previous implementation could we guarantee that all the "used" routing keys were always directed to the same client. On the other hand, by definition, the addition of a new client will have to steal something from others. For these reasons, in my opinion, the mixing of the modules does not involve any problem, but if instead it is to be considered problematic we can reason about an alternative solution.

**Comment by bert-laverman** [comment]
I'm not going to block this PR, but neither do I have anything useful to add at the mo. Withdrawing from list.

File: axonserver/src/main/java/io/axoniq/axonserver/message/command/hashing/ConsistentHashRoutingSelector.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
+package io.axoniq.axonserver.message.command.hashing;
+
+import io.axoniq.axonserver.message.command.RoutingSelector;
+import io.axoniq.axonserver.message.command.hashing.ConsistentHash.ConsistentHashMember;
+
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Function;
+
+/**
```
Comment (smcvb) [inline axonserver/src/main/java/io/axoniq/axonserver/message/command/hashing/ConsistentHashRoutingSelector.java]: Nit: Class level and constructor javadoc is missing.

File: axonserver/src/main/java/io/axoniq/axonserver/message/command/RoutingSelector.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
+package io.axoniq.axonserver.message.command;
+
+import java.util.Optional;
+
+/**
+ * Selects the best handler to dispatch the message to, chosen among registered handlers.
+ *
+ * @author Sara Pellegrini
+ * @since 4.3
```
Comment (smcvb) [inline axonserver/src/main/java/io/axoniq/axonserver/message/command/RoutingSelector.java]: Nit: The generic should be described in the javadoc

File: axonserver/src/test/java/io/axoniq/axonserver/message/command/LoadFactorBasedSelectorTest.java [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+        testSubject.register("15");
+        testSubject.register("30");
+        testSubject.register("50");
+        Map<String, AtomicInteger> counter = new HashMap<>();
+        int messages = 1_000_000;
+        for (int i = 0; i < messages; i++) {
+            String handler = testSubject.selectHandler(UUID.randomUUID().toString()).orElse(null);
+            counter.computeIfAbsent(handler, h -> new AtomicInteger()).incrementAndGet();
+        }
+        System.out.println(counter);
```
Comment (smcvb) [inline axonserver/src/test/java/io/axoniq/axonserver/message/command/LoadFactorBasedSelectorTest.java]: Think the `System.out.println` can be dropped.

File: axonserver/src/test/java/io/axoniq/axonserver/message/command/LoadFactorBasedSelectorTest.java [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+        counter.forEach((handler, count) -> assertEquals(Double.parseDouble(handler),
```
Comment (smcvb) [inline axonserver/src/test/java/io/axoniq/axonserver/message/command/LoadFactorBasedSelectorTest.java]: I do not think that this assertion is overly clear on face value. Would you mind extracting the `count.get() * 100d / messages` outcome to a named variable, just so others can immediately see without required deduction what's being asserted against?

File: axonserver/src/test/java/io/axoniq/axonserver/message/command/LoadFactorBasedSelectorTest.java [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+        testSubject.register("50");
+        Map<String, AtomicInteger> counter = new HashMap<>();
+        String handler = null;
+        String routingKey = UUID.randomUUID().toString();
+        for (int i = 0; i < 1_000_000; i++) {
+            handler = testSubject.selectHandler(routingKey).orElse(null);
+            counter.computeIfAbsent(handler, h -> new AtomicInteger()).incrementAndGet();
+        }
+        for (Map.Entry<String, AtomicInteger> entry : counter.entrySet()) {
+            if (entry.getKey().equals(handler)) {
```
Comment (smcvb) [inline axonserver/src/test/java/io/axoniq/axonserver/message/command/LoadFactorBasedSelectorTest.java]: Do you even get any other `handler` instances in the `counter` `Map` at all?

File: axonserver/src/main/java/io/axoniq/axonserver/message/command/CommandRegistrationCache.java [hunk 6]
```diff
@@ -138,25 +145,36 @@ private boolean provides(ClientIdentification client, String cmd) {
...
-        ConsistentHash hash = consistentHashPerContext.get(context);
-        if( hash == null) return null;
-        String client = hash
-                                                .getMember(routingKey, request.getName())
-                                                .map(ConsistentHash.ConsistentHashMember::getClient)
-                                                .orElse(null);
-        if( client == null ) return null;
+        String command = request.getName();
+        RoutingSelector<String> routingSelector = routingSelector(context, command);
+        String client = routingSelector.selectHandler(routingKey).orElse(null);
```
Comment (smcvb) [inline axonserver/src/main/java/io/axoniq/axonserver/message/command/CommandRegistrationCache.java]: Nit: you could change this line to:
```
return routingSelector.selectHandler(routingKey)
        .orElse(commandHandlersPerClientContext.get(new ClientIdentification(context, client)));
```

File: axonserver/src/main/java/io/axoniq/axonserver/message/command/CommandRegistrationCache.java [hunk 7]
```diff
@@ -138,25 +145,36 @@ private boolean provides(ClientIdentification client, String cmd) {
...
+        if (client == null) {
+            return null;
+        }
         return commandHandlersPerClientContext.get(new ClientIdentification(context, client));
```
Comment (smcvb) [inline axonserver/src/main/java/io/axoniq/axonserver/message/command/CommandRegistrationCache.java]: Nit: you could change this line to:
```
return routingSelector.selectHandler(routingKey)
        .orElse(commandHandlersPerClientContext.get(new ClientIdentification(context, client)));
```

Comment (saratry) [reply axonserver/src/main/java/io/axoniq/axonserver/message/command/CommandRegistrationCache.java]: client is a String, we want here the `CommandHandler`

File: axonserver/src/test/java/io/axoniq/axonserver/message/command/LoadFactorBasedSelectorTest.java (hunk 5 above)
Comment (saratry) [reply axonserver/src/test/java/io/axoniq/axonserver/message/command/LoadFactorBasedSelectorTest.java]: you are right, I'm going to refactor in order to check if the size is 1 and the content is the expected one.

File: axonserver/src/test/java/io/axoniq/axonserver/message/command/LoadFactorBasedSelectorTest.java (hunk 4 above)
Comment (saratry) [reply axonserver/src/test/java/io/axoniq/axonserver/message/command/LoadFactorBasedSelectorTest.java]: yes, you're right. I'm going to extract a variable named `percentage`

File: axonserver/src/test/java/io/axoniq/axonserver/message/command/LoadFactorBasedSelectorTest.java (hunk 3 above)
Comment (saratry) [reply axonserver/src/test/java/io/axoniq/axonserver/message/command/LoadFactorBasedSelectorTest.java]: yes

File: axonserver/src/main/java/io/axoniq/axonserver/message/command/RoutingSelector.java (hunk 2 above)
Comment (saratry) [reply axonserver/src/main/java/io/axoniq/axonserver/message/command/RoutingSelector.java]: Right

File: axonserver/src/test/java/io/axoniq/axonserver/message/command/LoadFactorBasedSelectorTest.java (hunk 4 above)
Comment (smcvb) [reply axonserver/src/test/java/io/axoniq/axonserver/message/command/LoadFactorBasedSelectorTest.java]: Thank you! :-)

File: axonserver/src/test/java/io/axoniq/axonserver/message/command/LoadFactorBasedSelectorTest.java (hunk 5 above)
Comment (smcvb) [reply axonserver/src/test/java/io/axoniq/axonserver/message/command/LoadFactorBasedSelectorTest.java]: Thank you! :-)

File: axonserver/src/main/java/io/axoniq/axonserver/message/command/CommandRegistrationCache.java (hunk 7 above)
Comment (smcvb) [reply axonserver/src/main/java/io/axoniq/axonserver/message/command/CommandRegistrationCache.java]: Aaah fair point, so a `map` operation should be added I guess.

File: axonserver/src/main/java/io/axoniq/axonserver/message/command/hashing/ConsistentHashRoutingSelector.java (hunk 1 above)
Comment (saratry) [reply axonserver/src/main/java/io/axoniq/axonserver/message/command/hashing/ConsistentHashRoutingSelector.java]: Done

File: axonserver/src/main/java/io/axoniq/axonserver/message/command/CommandRegistrationCache.java (hunk 7 above)
Comment (saratry) [reply axonserver/src/main/java/io/axoniq/axonserver/message/command/CommandRegistrationCache.java]: Done

**Pull request merged**